source/main.c

```diff
@@ -256,31 +357,65 @@ int main()
     if (load_fat("sd2", &__io_gcsd2)) goto load;
 
 load:
-    // Wait to exit while the d-pad down direction is held.
-    while (all_buttons_held & PAD_BUTTON_DOWN)
-    {
-        VIDEO_WaitVSync();
-        PAD_ScanPads();
-        all_buttons_held = (
-            PAD_ButtonsHeld(PAD_CHAN0) |
-            PAD_ButtonsHeld(PAD_CHAN1) |
-            PAD_ButtonsHeld(PAD_CHAN2) |
-            PAD_ButtonsHeld(PAD_CHAN3)
-        );
-    }
-
```

Comment:
Move the whole `if (dol_argc)` out of `if (dol)` and you won't need to extract this bit into a function.

Reply:
Yea this whole block feels awkward. Although we know that *currently* there is no way `dol_argc` can be set without `dol`, it felt disconnected to have logic that reads as if `dol_argc` could be processed even if `dol` is not set. Feels weird to atomize the loading of the args and the loading of the DOL as the former doesn't make sense without the later.

I did it this way because I assumed there two distinct exit paths (with and without a DOL) which could not converge because of `SYS_ResetSystem`/`SYS_SwitchFiber`. I don't know exactly how these interplay.

Given that `SYS_ResetSystem` is not directly related to the loading of the DOL, we could move it the end if possible. Something like:

```c
  bool should_reset = false;
  if (!dol) goto exit;

  if (dol_argc)
  {
    // ... load args
  }
  // ... load dol
  should_reset = true;

exit:
  // wait for d-pad down to be released
  if (should_reset)
  {
    SYS_ResetSystem(SYS_SHUTDOWN, 0, FALSE);
  }
  return 0;
```

Reply:
I'm pretty sure this particular call is to shut down the libogc "OS" before jumping into the relocation stub. @Extrems will have to clarify this.

Returning from main otherwise returns to libogc's startup code which will do its own cleanup and then a hard reset (which is how we reboot into the original IPL, in fact it's the only way because you can't reset and reactivate the bootrom decryption block from software). Technically we could read, decrypt and execute the original IPL like a regular DOL, but that's really unnecessary.

Reply:
> This is only possible if it's OK to have to call SYS_ResetSystem after SYS_SwitchFiber and after a delay.

That absolutely isn't ok. The operating system must be shutdown before jumping into the relocation stub.

Reply:
I figured something like that was the case.

Working back from there, we do indeed have 2 separate and non-converging exit paths. We can't wait for d-down before the split as one of the paths does some logging. And we can't wait after because one of the paths never returns.

The solution is to either mix the paths or allow each path to determine when to wait. I went with the later. If I were to refactor I would do something like:

```c
if (!dol)
{
  delay_exit();
  return 0;
}

if (dol_argc)
{
  // ... load args
}
// ... load dol

delay_exit();

SYS_ResetSystem(...);
SYS_SwitchFiber(...);
return 0;
```

Reply:
Pushed the refactor described above.